### PR TITLE
hardening/errorAttrsWithoutValue

### PR DESCRIPTION
**Pull request by rbl219**
Resolved issue   #1164

Clearer error on "PUT /v2/entities/<entityId>/attrs/<atrtrName>" without 'value' in payload

### CODE REVIEW

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 1]
```diff
@@ -240,6 +240,7 @@ Error list (HTTP response code in paranthesis):
 * LengthRequired (411). Zero/No Content-Length in PUT/POST/PATCH request
 * RequestEntityTooLarge (413). Payload too large
 * UnsupportedMediaType (415). Request content type is not supported.
+* InvalidModification (422). Request payload is missing some piece of information.
```
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Maybe is better English to say?

"Request payload misses some piece of information"

File: doc/apiary/v2/fiware-ngsiv2-reference.apib [hunk 2]
```diff
@@ -240,6 +240,7 @@ Error list (HTTP response code in paranthesis):
 * LengthRequired (411). Zero/No Content-Length in PUT/POST/PATCH request
```
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: CNR file entry corresponding to this PR is missing.

File: src/lib/mongoBackend/MongoCommonUpdate.cpp [hunk 3]
```diff
@@ -2239,7 +2240,7 @@ void processContextElement
       {
         ContextAttribute* ca = new ContextAttribute(aP);
 
-        buildGeneralErrorResponse(ceP, ca, responseP, SccInvalidParameter,
+        buildGeneralErrorResponse(ceP, ca, responseP, SccInvalidModification,
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Interesting point to think about it... :)

This change introduces a slight backward compability break, as we are returning 422 in places where 472 was returned previously. However, in this case we are moving from a general error (invalid parameter, in general) to a more specific error (invalid modification), so I think is actually an improvement.

We could avoid this and preserve previous NGSIv1 behaviour using the "if (apiVersion == v2)" technique already used in other PRs. However, in this particular case I think we should leave the modification as @rbl219 suggest.

Opinions?

File: test/functionalTest/cases/0000_ngsi10_update/update_sequence_empty_value_in_context_value.test [hunk 4]
```diff
@@ -72,8 +72,8 @@ Date: REGEX(.*)
         </contextAttributeList>
       </contextElement>
       <statusCode>
-        <code>472</code>
-        <reasonPhrase>request parameter is invalid/not allowed</reasonPhrase>
+        <code>422</code>
+        <reasonPhrase>Invalid Modification</reasonPhrase>
         <details>action: APPEND - entity: [OfficeRoom, Room, false] - offending attribute: pressure - empty attribute not allowed in APPEND or UPDATE</details>
       </statusCode>
     </contextElementResponse>
```
Comment (fgalan) [inline test/functionalTest/cases/0000_ngsi10_update/update_sequence_empty_value_in_context_value.test]: Could you include a test harness to test the fixing in the NGSIv2 API, pls? Contact me if you have doubt about it :)

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 2 above)
Comment (rbl219) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fix in commit ed18967

File: test/functionalTest/cases/0000_ngsi10_update/update_sequence_empty_value_in_context_value.test (hunk 4 above)
Comment (rbl219) [inline test/functionalTest/cases/0000_ngsi10_update/update_sequence_empty_value_in_context_value.test]: ok, I will include new test harness

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: I would vote for not breaking backward compatibility ...

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 1 above)
Comment (kzangeli) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Better would be to turn it around:

Some piece of information (is) missing in payload

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 3 above)
Comment (fgalan) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: After a skype discussion, at the end we will keep the code as it is now, due to:

* Backward compabilitiy is really really small
* It involves an improvement
* Keep backward compability infolves extra work
* Kepp backward compability involves "dirty" code due to the "if (apiVersion == v2)"  exceptions.

However, let's make this explicit in the CNR, with an additiona entry like this:

"Fix: chaned error code from 472 (Invalid Parameter) changed to 422 (Invalid Moficiation) in NGSIv1 in the case of missing attribute value (#1164)"

File: test/functionalTest/cases/0000_ngsi10_update/update_sequence_empty_value_in_context_value.test (hunk 4 above)
Comment (rbl219) [inline test/functionalTest/cases/0000_ngsi10_update/update_sequence_empty_value_in_context_value.test]: Fix in commit 6d6b824

File: src/lib/mongoBackend/MongoCommonUpdate.cpp (hunk 3 above)
Comment (rbl219) [inline src/lib/mongoBackend/MongoCommonUpdate.cpp]: Fix in commit 14fa438

File: doc/apiary/v2/fiware-ngsiv2-reference.apib (hunk 1 above)
Comment (fgalan) [inline doc/apiary/v2/fiware-ngsiv2-reference.apib]: Fixed.

**Comment by fgalan** [comment]
LGTM (but unable to merge yet due develop need to be merged...)

**Comment by kzangeli** [comment]
LGTM, after pending merge ...

**Comment by rbl219** [comment]
Merged in commit ff5521c

**Pull request merged**